src/main/java/io/temporal/samples/hello/HelloDetachedCancellationScope.java

```diff
@@ -0,0 +1,226 @@
+      // Simulate some long-running activity so we can cancel workflow before it completes
+      for (int i = 0; i < Integer.MAX_VALUE; i++) {
+        try {
+          Thread.sleep(200);
+        } catch (InterruptedException e) {
+          // Rethrow the exception as runtime one
+          throw Activity.wrap(e);
+        }
+        Activity.getExecutionContext().heartbeat(i);
+      }
+      return "unreachable";
+    }
+
+    @Override
+    public String sayGoodBye(String name) {
+      return "Goodbye " + name + "!";
+    }
+  }
+
+  /** Define the workflow implementation which implements our getGreeting workflow method. */
+  public static class GreetingWorkflowImpl implements GreetingWorkflow {
+    private String greeting;
+
+    private final GreetingActivities activities =
+        Workflow.newActivityStub(
+            GreetingActivities.class,
+            ActivityOptions.newBuilder()
+                .setStartToCloseTimeout(Duration.ofSeconds(10))
+                .setCancellationType(ActivityCancellationType.WAIT_CANCELLATION_COMPLETED)
+                .build());
+
+    @Override
+    public String getGreeting(String name) {
+      try {
+        this.greeting = activities.sayHello(name);
+        return greeting;
+      } catch (ActivityFailure af) {
+        // Create a CancellationScope that is not linked to a parent scope
+        // This can be used to do cleanup code after workflow has been cancelled
+        CancellationScope detached =
+            Workflow.newDetachedCancellationScope(() -> greeting = activities.sayGoodBye(name));
+        detached.run();
+        throw af;
+      }
+    }
+
+    @Override
+    public String queryGreeting() {
+      return greeting;
+    }
+  }
+
+  public static void main(String[] args) throws InterruptedException {
+
+    // Define the workflow service.
+    WorkflowServiceStubs service = WorkflowServiceStubs.newInstance();
+
+    /**
+     * Define the workflow client. It is a Temporal service client used to start, signal, and query
+     * workflows
+     */
+    WorkflowClient client = WorkflowClient.newInstance(service);
+
+    /**
+     * Define the workflow factory. It is used to create workflow workers for a specific task queue.
+     */
+    WorkerFactory factory = WorkerFactory.newInstance(client);
+
+    /**
+     * Define the workflow worker. Workflow workers listen to a defined task queue and process
+     * workflows and activities.
+     */
+    Worker worker = factory.newWorker(TASK_QUEUE);
+
+    /**
+     * Register our workflow implementation with the worker. Workflow implementations must be known
+     * to the worker at runtime in order to dispatch workflow tasks.
+     */
+    worker.registerWorkflowImplementationTypes(GreetingWorkflowImpl.class);
+
+    /**
+     * Register our workflow activity implementation with the worker. Since workflow activities are
+     * stateless and thread-safe, we need to register a shared instance.
```

Comment:
```suggestion
     * Register our Activity Types with the Worker. Since Activities are
     * stateless and thread-safe, the Activity Type is a shared instance.
```

Reply:
Global change - recorded